### ISSUE TITLE
Examples of EnzymeRegister*Handler C API

### ISSUE THREAD

**Issue by oursland** [wsmoses/Enzyme#324]
I am working on a project in which custom derivatives for functions may be computed more quickly than with numerical differentiation.

Since [0563ba4746a3d29298f01afee9372cd213e2fee8](https://github.com/wsmoses/Enzyme/commit/0563ba4746a3d29298f01afee9372cd213e2fee8), functions for [EnzymeRegister*Handler](https://github.com/wsmoses/Enzyme/commit/0563ba4746a3d29298f01afee9372cd213e2fee8#diff-26a88d10ea20adcfdbf4bb74547f89964b4e19ae9ca6b4508b3063b101f58732R241) have been present in the C API. These functions appear to me to be a means to register custom function derivatives.

Are there any examples on how one implements custom derivatives for functions?

**Comment by wsmoses** [comment]
Presently the only use of programmatic derivative registration is within Enzyme.jl, see here: https://github.com/wsmoses/Enzyme.jl/blob/d9113ac31d6cb94251cd9a44af844423877b1ccc/src/compiler.jl#L973 This is especially useful for JITs and other scenarios with dynamically generated functions.

Individual custom derivatives can also be registered individually by creating either a magic global containing `enzyme_register_gradient` in its name with the original function, the custom augmented forward pass, and a corresponding reverse pass. A (bit complicated) example of that can be seen in the CUDA guide here: https://enzyme.mit.edu/getting_started/CUDAGuide/#heterogeneous-ad (scroll to `__enzyme_register_gradient_kern` ) and here: https://github.com/wsmoses/Enzyme/issues/142#issuecomment-781895318.

This is definitely lacking in documentation, feel free to make a PR to the website branch!

Out of curiosity, what is your use case? Are you looking more for programatic registration or a simple custom derivative?

**Comment by oursland** [comment]
I am writing some simulation code intended for small embedded systems in C, so the use case is more of a simple custom derivative.

The suggestions on #142 have been very helpful, however I did have to make some modifications to get the recommended code sample to compile and it fails to do so when not explicitly providing custom derivatives (Eigen::exp<MatrixXd>() and Eigen::inverse<MatrixXd>() appear to trigger undesirable behavior).

I have attempted to create custom derivatives on the basic getting started example and cannot get Enzyme to use the custom augment and gradient functions.  As this differs significantly from the original title and description of the current issue, I have created issue #327 with more details.
